worker/caasapplicationprovisioner/application.go

```diff
@@ -97,20 +97,17 @@ func (a *appWorker) Wait() error {
 }
 
 func (a *appWorker) loop() error {
-	charmURL, err := a.facade.ApplicationCharmURL(a.name)
+	// If charm is (now) a v1 charm, exit the worker.
+	format, err := a.charmFormat()
 	if errors.IsNotFound(err) {
 		a.logger.Debugf("application %q removed", a.name)
 		return nil
 	} else if err != nil {
-		return errors.Annotatef(err, "failed to get charm url for application")
-	}
-	charmInfo, err := a.facade.CharmInfo(charmURL.String())
-	if err != nil {
-		return errors.Annotatef(err, "failed to get application charm deployment metadata for %q", a.name)
+		return errors.Trace(err)
 	}
-	if charmInfo == nil ||
-		corecharm.Format(charmInfo.Charm()) < corecharm.FormatV2 {
-		return errors.Errorf("charm version 2 or greater required")
+	if format < corecharm.FormatV2 {
+		a.logger.Debugf("application %q v2 worker got v1 charm event, stopping", a.name)
+		return nil
```

Comment:
If we were to disallow downgrades from v1 to v2, we could do this in the parent worker, ie avoid starting this worker only to stop it straight away

Reply:
Done.